Post-Training Quantization of MobileNet v2 PyTorch Model

alexsu52 opened the pull request:
### Changes

Introducing the Post-Training Quantization of MobileNet v2 PyTorch Model example

### Reason for changes

Moving the example from the experimental module

### Related tickets

97478

### Tests

TBD

AlexKoff88 commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,136 @@
+
+    print('Benchmark INT8 model (OpenVINO)')
+    int8_benchmark_results = ov_benchmark(int8_ir_path)
+    print(int8_benchmark_results)
```
A performance boost would be reasonable since you report an accuracy difference

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,136 @@
+import subprocess
+import nncf
+import numpy as np
+import openvino
```
It doesn't work. I think you should use
```python
import openvino.runtime as ov
```

l-bat commented on examples/post_training_quantization/torch/mobilenet_v2/README.md:
```diff
@@ -0,0 +1,26 @@
+# Post-Training Quantization of MobileNet v2 PyTorch Model
+This sample demonstrates how to use Post-Training Quantization API from Neural Network Compression Framework (NNCF) to quantize PyTorch models on the example of [MobileNet v2](https://huggingface.co/AlexKoff88/mobilenet_v2_food101)quatization, pretrained on [Food101](https://pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html) dataset.
```
```suggestion
This sample demonstrates how to use Post-Training Quantization API from Neural Network Compression Framework (NNCF) to quantize PyTorch models on the example of [MobileNet v2](https://huggingface.co/AlexKoff88/mobilenet_v2_food101) quantization, pretrained on [Food101](https://pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html) dataset.
```

l-bat commented on examples/post_training_quantization/torch/mobilenet_v2/README.md:
```diff
@@ -0,0 +1,26 @@
+
+# Run Sample
+It's pretty simple. The sample does not require additional preparation. It will do the preparation itself, such as loading the dataset and model, etc.
+> **NOTE**: If you have already downloaded the [Food101](https://pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html) dataset to the disk, please provide the path to the daatset to avoid re-downloading.
```
```suggestion
> **NOTE**: If you have already downloaded the [Food101](https://pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html) dataset to the disk, please provide the path to the dataset to avoid re-downloading.
```

aiarkinx commented:
run onnx pre-commit tests

aiarkinx commented:
run openvino pre-commit tests

vshampor commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+FOOD101_CLASSES = 101
+ROOT = Path(__file__).parent.resolve()
+DATASET_PATH = Path(sys.argv[1]) if len(sys.argv) > 1 else ROOT / 'dataset'
+CHECKPOINT_URL = 'https://huggingface.co/AlexKoff88/mobilenet_v2_food101/resolve/main/pytorch_model.bin'
```
We probably should have it at S3 instead.

alexsu52 replied:
I will do

l-bat commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+print('[7/7] Report:')
+print(f'Accuracy drop: {fp32_top1 - int8_top1:.3f}')
+print(f'Compression rate: {fp32_model_size / int8_model_size:.3f}')
+# https://docs.openvino.ai/2018_R5/_docs_IE_DG_Intro_to_Performance.html
```
According to `examples/post_training_quantization/torch/mobilenet_v2/requirements.txt` `openvino-dev>=2022.2.0` is required, but the link refers to `OpenVINO 2018.5`. I think the information may be outdated or confuse users.

alexsu52 replied:
Done

helena-intel reviewed (commented):
This is a great demo! I found it easy to follow and it highlights that the quantization itself is only two lines of code. This concise report at the end with all the relevant information is awesome: 
![image](https://user-images.githubusercontent.com/77325899/203953545-7dc15efd-feef-48dd-aa4b-44bb199458df.png)

I noticed that INT8 is about 3 times as small as FP32, not 4 times. That was a bit unexpected to me. 

Some comments about nice-to-haves:
- I would love to see docstrings and possibly type hints for the functions. It may seem overkill for such simple functions but I think it is really useful
- The dataset will be slow to download for users with slower internet, and you need 10GB of space to download and unpack it. It would be great if this could work with a smaller dataset.

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+###########################################################################
+# Create a PyTorch model and dataset
+###########################################################################
+normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
```
What do you think about using
```python
def main():
    ...

if __name__ == '__main__':
    main()
```
https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md#311-main

helena-intel replied:
Personally I prefer the current approach. IMO these main() functions are C-isms. They have their use in scripts that can be used standalone and imported, but for a simple demo, I really like the way it is now.

alexsu52 replied:
I agree with @helena-intel

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+    return accuracy_score(predictions, references)
+
+
+def ov_convert(model, args):
```
My suggestion is to use "actions" in the methods names. I mean,
`ov_convert()` --> `convert_to_openvino_model()`
`ov_benchmark()` --> `run_benchmark()`
`ov_model_size()` --> `get_model_size()`

or something like this. What do you think?

alexsu52 replied:
Done

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def ov_convert(model, args):
+    onnx_model_path = f'{tempfile.gettempdir()}/model.onnx'
```
I think it is better to use such approach
```python
with tempfile.TemporaryDirectory(dir=tempfile.gettempdir()) as temp_dir:
    ...
```
because `temp_dir` will be removed after exiting from the `with` statement.

alexsu52 replied:
The function was removed.

andrey-churkin replied:
There is another place where it can be fixed

https://github.com/alexsu52/nncf/blob/9acf21d6cedb15487e1eaf79731039f630082995/examples/post_training_quantization/torch/mobilenet_v2/main.py#L70-L72

KodiaqQ commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+val_dataset = datasets.Food101(
+    root=DATASET_PATH,
+    split = 'test', 
+    transform = transforms.Compose([
```
It would be great to add a link to a description of why we use these particular transformations. It looks like `ImageNet` transformations, isn't it?

kshpv commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+CHECKPOINT_URL = 'https://huggingface.co/AlexKoff88/mobilenet_v2_food101/resolve/main/pytorch_model.bin'
+
+
+def fix_names(state_dict):
```
This is not clear why we should call this function. Maybe it is better to update the checkpoint itself and completely remove this function?

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+torchvision>=0.12, <0.14
+tqdm
+sklearn
```
> deprecated sklearn package, use scikit-learn instead
https://pypi.org/project/sklearn/

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/README.md:
````diff
@@ -0,0 +1,26 @@
+# Run Sample
+It's pretty simple. The sample does not require additional preparation. It will do the preparation itself, such as loading the dataset and model, etc.
+> **NOTE**: If you have already downloaded the [Food101](https://pytorch.org/vision/stable/generated/torchvision.datasets.Food101.html) dataset to the disk, please provide the path to the dataset to avoid re-downloading.
+```
+python main.py [PATH_TO_DATASET]
+```
````
This logic was not implemented.

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+import openvino.runtime as ov
+import torch
+from sklearn.metrics import accuracy_score
+from openvino.offline_transformations import compress_quantize_weights_transformation
```
I got the following warning:

> DeprecationWarning: compress_quantize_weights_transformation is deprecated and will be removed in version 2023.1. The module is private and following namespace `offline_transformations` will be removed in the future.

Should we care about it?

alexsu52 commented:
> 

I have replaced food101 to imagenette dataset.

alexsu52 commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+###########################################################################
+
+'''
+Transformation function transforms a data item into model input.
```
Yes, you are right. the calibration dataset should contain only the input data. I added some comment what the calibration dataset is.

alexsu52 commented on examples/post_training_quantization/torch/mobilenet_v2/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+torchvision>=0.12, <0.14
```
OpenVINO 2022.2 does not have full support for torch 1.13, in practice some low precision models show poor performance. When OpenVINO 2022.3 is released, I will remove the top border of torchvision => `torchvision>=0.12`

alexsu52 commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def load_checkpoint(model):  
+    checkpoint = torch.hub.load_state_dict_from_url(CHECKPOINT_URL, progress=False)
```
Done

alexsu52 commented:
> This is a great demo! I found it easy to follow and it highlights that the quantization itself is only two lines of code. This concise report at the end with all the relevant information is awesome: ![image](https://user-images.githubusercontent.com/77325899/203953545-7dc15efd-feef-48dd-aa4b-44bb199458df.png)
> 
> I noticed that INT8 is about 3 times as small as FP32, not 4 times. That was a bit unexpected to me.
> 
> Some comments about nice-to-haves:
> 
>     * I would love to see docstrings and possibly type hints for the functions. It may seem overkill for such simple functions but I think it is really useful
> 
>     * The dataset will be slow to download for users with slower internet, and you need 10GB of space to download and unpack it. It would be great if this could work with a smaller dataset.

Thanks for your comments! Your feedback is very helpful!

* OpenVINO IR has xml (model graph) and bin (model weights) files. The compression ratio is calculated using the model size as a summary of the xml size and the bin size. The quantization algorithm only compresses the model weights, not the model graph. The model graph is slightly larger because FQ nodes have been added to the model graph. So the compression factor will be < 4. Anyway, I have added an extra output with xml size and bin size.
* I've added type hints for the functions. I hope the function names are clear and don't require docstrings.
* I've changed food101 to imagenette dataset

AlexKoff88 commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,189 @@
+ROOT = Path(__file__).parent.resolve()
+CHECKPOINT_URL = 'https://huggingface.co/alexsu52/mobilenet_v2_imagenette/resolve/main/pytorch_model.bin'
+DATASET_URL = 'https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-320.tgz'
+DATASET_PATH = '~/.nncf'
```
`~/.cache/nncf/datasets/` should be more appropriate, IMO. At least, `~/.cache/...` is the folder that is used by default by may other libraries, including torch.

alexsu52 replied:
Done

AlexKoff88 commented on examples/post_training_quantization/torch/mobilenet_v2/main.py:
```diff
@@ -0,0 +1,189 @@
+    predictions = []
+    references = []
+
+    output_name = model.outputs[0]
```
`model.outputs[0]` is not the name but object so it is better to rename `output_name` to `output`

alexsu52 replied:
Done

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/README.md:
```diff
@@ -0,0 +1,25 @@
+    - Performance speed up of the quantized model (INT8)
+
+# Install requirements
+At this point it is assumed that you have already installed nncf. You can find information on installation nncf [here](https://github.com/openvinotoolkit/nncf#user-content-installation).
```
```suggestion
At this point it is assumed that you have already installed NNCF. You can find information on installation NNCF [here](https://github.com/openvinotoolkit/nncf#user-content-installation).
```

alexsu52 replied:
Done

andrey-churkin commented on examples/post_training_quantization/torch/mobilenet_v2/README.md:
````diff
@@ -0,0 +1,25 @@
+pip install -r requirements.txt
+```
+
+# Run Sample
````
I think it is better to use `Example` instead of `Sample` because our folder has the name `examples`.

alexsu52 replied:
Done
